include/Sema/ConstraintSystem.hpp

```diff
@@ -0,0 +1,70 @@
+#ifndef GLU_SEMA_CONSTRAINT_SYSTEM_HPP
+#define GLU_SEMA_CONSTRAINT_SYSTEM_HPP
+
+#include "Constraint.hpp"
+#include "ScopeTable.hpp"
+
+#include <llvm/ADT/DenseMap.h>
+#include <llvm/ADT/ilist.h>
+
+namespace glu::sema {
+
+class ConstraintSystem {
+    ScopeTable *_scopeTable; // The scope table for the current context
+    std::vector<glu::types::TypeVariableTy *>
+        _typeVariables; // List of type variables
+    llvm::BumpPtrAllocator _allocator; // Allocator for memory management
+    llvm::ilist<Constraint> _constraints; // List of constraints
```

Comment:
Constraints aren’t an intrusive list? We really only have one constraint which is a conjunction of all the others. Or a list of constraints directly. But it’s not intrusive, there’s no prev or next in class Constraint